// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package bap;

import "buf/validate/validate.proto";
import "benchmarking/proto/common/workload_action.proto";
import "benchmarking/proto/common/metric.proto";
import "benchmarking/proto/common/workflow_type.proto"; 

// Defines the group assignment for A/B testing scenarios.
enum AbTestGroup {
  // Default: Standard single execution.
  AB_TEST_GROUP_UNSPECIFIED = 0;
  // The control group (baseline/reference).
  BASELINE = 1;
  // The treatment group (experiment/candidate).
  EXPERIMENT = 2;
}

// Represents a single, fully resolved job in the GitHub Actions matrix.
message BenchmarkJob {
  // REQUIRED: Unique identifier for this specific combination of benchmark configuration
  // and environment configuration (e.g. "resnet50_gpu_v100").
  string config_id = 1 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: The workflow type associated with this run.
  bap.common.WorkflowType workflow_type = 2 [(buf.validate.field).enum = {not_in: [0]}];
  
  // REQUIRED: The human-readable name of the benchmark.
  string benchmark_name = 3 [(buf.validate.field).string.min_len = 1];
  
  // REQUIRED: Description of the benchmark's purpose.
  string description = 4 [(buf.validate.field).string.min_len = 1];
  
  // REQUIRED: Owner or team responsible for the benchmark.
  string owner = 5 [(buf.validate.field).string.min_len = 1];
  
  // Optional: GitHub labels for filtering or triggering workflows.
  repeated string github_labels = 6; 
  
  // REQUIRED: The GitHub Actions runner label (e.g. "linux-x86-n2-32").
  string runner_label = 7 [(buf.validate.field).string.min_len = 1];
  
  // REQUIRED: The container image to execute the workload inside.
  string container_image = 8 [(buf.validate.field).string.min_len = 1];  
  
  // REQUIRED: Configuration for the workload executor action.
  bap.common.WorkloadAction workload = 9 [(buf.validate.field).required = true];
  
  // Optional: List of metrics to track.
  repeated bap.common.MetricSpec metrics = 10;

  // OPTIONAL: The A/B test group this job belongs to.
  // Defaults to AB_TEST_GROUP_UNSPECIFIED (standard run).
  AbTestGroup ab_test_group = 11;

  // OPTIONAL: The specific Git reference (SHA, branch, tag) to checkout for this job.
  // Generally required if ab_test_group is BASELINE or EXPERIMENT.
  string checkout_ref = 12;
}
